server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceInput.java

```diff
@@ -37,10 +38,10 @@ private static Set<ShardRouting> getIgnoredShardsWithDiscardedAllocationStatus(R
             .unassigned()
             .ignored()
             .stream()
-            .map(
-                shardRouting -> shardRouting.updateUnassigned(
-                    discardAllocationStatus(shardRouting.unassignedInfo()),
-                    shardRouting.recoverySource()
+            .flatMap(
+                shardRouting -> Stream.of(
+                    shardRouting,
+                    shardRouting.updateUnassigned(discardAllocationStatus(shardRouting.unassignedInfo()), shardRouting.recoverySource())
                 )
```

Comment:
Related to https://github.com/elastic/elasticsearch/pull/90600

In some cases the allocation status is coming from GatewayAllocator and need to be reseted to `NO_ATTEMPT` to be able to find the ignored shard in others (NO_VALID_SHARD_COPY) it is kept as is.

I wonder if this indicates that we want to have some other way of identifying ignored shards bypassing its entire state and unassigned info?